### PR TITLE
feat: convert central stores to use injection systems

### PR DESCRIPTION
**Pull request by kentquirk**
Changes enabling injection of the central store.

Note that this is against the dynamic scaling branch.

### CODE REVIEW

**Comment by kentquirk** [comment]
The errors here are because config values aren't set up in the metadata yet; I'd like to do that in a separate PR; this one is already big enough.

File: config/file_config.go [hunk 1]
```diff
@@ -213,6 +214,16 @@ type CollectionConfig struct {
...
 }
 
+type SmartWrapperOptions struct {
+	BasicStoreType    string   `yaml:"Type" default:"local"`
+	SpanChannelSize   int      `yaml:"SpanChannelSize" default:"100"`
+	StateTicker       Duration `yaml:"StateTicker" default:"1s"`
+	SendDelay         Duration `yaml:"SendDelay" default:"2s"`
+	TraceTimeout      Duration `yaml:"TraceTimeout" default:"60s"`
+	DecisionTimeout   Duration `yaml:"DecisionTimeout" default:"3s"`
+	MaxTraceRetention Duration `yaml:"MaxTraceRetention" default:"24h"`
```
Comment (VinozzZ) [inline config/file_config.go:224]: I think `MacTraceRetention` is only used in redis store?

File: centralstore/localremotestore.go [hunk 2]
```diff
@@ -6,77 +6,39 @@ import (
...
 
-	if err != nil {
-		// TODO: handle this better
-		panic(err)
-	}
-	lrs.decisionCache = decisionCache
+func (lrs *LocalRemoteStore) Start() error {
+	lrs.states = make(map[CentralTraceState]statusMap)
+	lrs.traces = make(map[string]*CentralTrace)
+	lrs.Clock = clockwork.NewRealClock()
```
[the remark this answers is not in the document]
Comment (kentquirk) [reply centralstore/localremotestore.go:41]: Oh, I missed this -- I have injected one, I'll fix.

File: centralstore/redis_store.go [hunk 3]
```diff
@@ -983,31 +983,31 @@ const stateChangeScript = `
 	    if (currentState == nil or currentState == false) then
 	 	  currentState = previousState
 	    end
-	
+
```
[the remark this answers is not in the document]
Comment (kentquirk) [reply centralstore/redis_store.go:986]: It's coming from someone (you?) who doesn't have "trim trailing whitespace" selected in their editor.

File: centralstore/smartwrapper.go [hunk 4]
```diff
@@ -15,58 +16,84 @@ type statusMap map[string]*CentralTraceStatus
...
 // NewSmartWrapper creates a new SmartWrapper.
-func NewSmartWrapper(options SmartWrapperOptions, basic BasicStorer, tracer trace.Tracer) *SmartWrapper {
-	w := &SmartWrapper{
-		basicStore:     basic,
-		stopped:        make(chan struct{}),
-		spanChan:       make(chan *CentralSpan, options.SpanChannelSize),
-		done:           make(chan struct{}),
-		doneProcessing: make(chan struct{}),
-		tracer:         tracer,
+// func NewSmartWrapper(options SmartWrapperOptions, basic BasicStorer, tracer trace.Tracer) *SmartWrapper {
```
[the remark this answers is not in the document]
Comment (kentquirk) [reply centralstore/smartwrapper.go:40]: No, I just had a file still open when I committed; will fix.

File: config/file_config.go (hunk 1 above)
Comment (kentquirk) [reply config/file_config.go:224]: I was thinking it's potentially useful in some future store. Are you doing additional config work? If so, you could move it. I don't feel strongly, though.

Comment (VinozzZ) [reply config/file_config.go:224]: I see. That makes sense. I was working on adding additional Config so I moved it out. I will play with it more in my PR for making the redis to work with the injection framework. We can leave it as it is for this PR

File: config.md [hunk 5]
```diff
@@ -969,7 +969,7 @@ Enabled specifies whether the gRPC server is enabled.
 If `false`, then the gRPC server is not started and no gRPC traffic is accepted.
 
 - Not eligible for live reload.
-- Type: `defaulttrue`
+- Type: `bool`
```
Comment (VinozzZ) [inline config.md:972]: Is this change intended?

**Review by VinozzZ — APPROVED** [review]
LGTM. I just left one question about a change from merging with main

File: centralstore/smartwrapper.go [hunk 6]
```diff
@@ -15,58 +15,59 @@ type statusMap map[string]*CentralTraceStatus
...
+	go w.manageStates(context.Background(), opts)
 
-	return w
+	return nil
 }
 
-func (w *SmartWrapper) Stop() {
+func (w *SmartWrapper) Stop() error {
 	fmt.Println("stopping SmartWrapper")
+	w.closing = true
```
Comment (VinozzZ) [inline centralstore/smartwrapper.go:62]: Is `closing` being used anywhere? I can't seem to find it

Comment (kentquirk) [reply centralstore/smartwrapper.go]: It was an experiment that failed; good catch.

**Review by VinozzZ — APPROVED** [review]
